File: PROJECTS/BASICS/transformation.py
# ROBOTICS: MODULE TO APPLY A GENERAL TRANSFORMATION HOMOGENEOUS MATRIX
# Santiago Garcia Arango

import numpy as np
import math

class Transformation:
    """
    TRANSFORMATION MATRICES FOR ROBOTICS. Santiago Garcia Arango.
    angle_x = rotation of {B} with respect of {A} in "x" axis.
    angle_y = rotation of {B} with respect of {A} in "y" axis.
    angle_z = rotation of {B} with respect of {A} in "z" axis.
    vector_AB = vector from origin of {A} to origin of {B}.
    vector_BP = vector from origin of {B} to point "P".
    """
    def __init__(self, angle_x, angle_y, angle_z, vector_AB, vector_BP):
        # Add extra "1" to vector_AB
        vector_AB.append(1)
        self.vector_AB = np.transpose([vector_AB])

        # Add extra "1" to vector_BP
        vector_BP.append(1)
        self.vector_BP = np.transpose([vector_BP])

        # Initialize rotation matrices in zero
        self.RX = np.eye(3)
        self.RY = np.eye(3)
        self.RZ = np.eye(3)

        # Change the rotation matrices based on inputs
        self.create_rotation_matrix("x", angle_x)
        self.create_rotation_matrix("y", angle_y)
        self.create_rotation_matrix("z", angle_z)

        print("\nRX:\n", self.RX)
        print("\nRY:\n", self.RY)
        print("\nRZ:\n", self.RZ)

        # Create homogeneous transformation matrix
        self.create_transformation_matriz()

        # Find vector_AP (from main coordinate system to point B)
        self.vector_AP = np.dot(self.TM, self.vector_BP)


    def create_rotation_matrix(self, axis, angle):
        angle = math.radians(angle)  # Convert from degrees to radians

        if axis == "x":
            self.RX = np.array([
                [1, 0, 0],
                [0, math.cos(angle), -math.sin(angle)],
                [0, math.sin(angle), math.cos(angle)]
            ])

        if axis == "y":
            self.RY = np.array([
                [math.cos(angle), 0, math.sin(angle)],
                [0, 1, 0],
                [-math.sin(angle), 0, math.cos(angle)]
            ])

        if axis == "z":
            self.RZ = np.array([
                [math.cos(angle), -math.sin(angle), 0],
                [math.sin(angle), math.cos(angle), 0],
                [0, 0, 1]
            ])
    
    def create_transformation_matriz(self):
        R_ZYX = np.dot(np.dot(self.RZ, self.RY), self.RX)
        
        print("\nR_ZYX\n", R_ZYX)

        extra_zeros_perspective = np.array([[0, 0, 0]])
        self.TM = np.concatenate((R_ZYX, extra_zeros_perspective))
        self.TM = np.concatenate((self.TM, self.vector_AB), axis=1)

        print("\nTM:\n", self.TM)


# TESTS
if __name__ == "__main__":
    test = Transformation(0, 0, 30, [10, 5, 0], [3, 7, 0])
    print("\nvector_AP:\n", test.vector_AP)

File: PROJECTS/TRANSFORMATIONS/transformation.m
% GENERAL FUNCTION TO GET THE TRANSFORMATION MATRIX
% SANTIAGO GARCIA ARANGO
% Input analysis:
% vector_K: vector that will be the rotation axis. [kx, ky, kz] (1 x 3)
% angle: angle that will be rotated on vector_k. [entry syms or radians]
% vector_AB: vector from A org to B org. [ABx, ABy, ABz] (1 x 3)


function [TM] = transformation(vector_K, vector_AB, angle)
% TRANSFORMATION--> Return transformation matrix from spceific parameters

if nargin < 3
    syms theta
    vector_K = [0.7070, 0.7070, 0];
    vector_AB = [0, 0, 0];
    angle = deg2rad(30);
    fprintf("\n \n \n \n *ENTER ALL PARAMETERS TO FUNCTION* \n \n \n \n");
end


% Add extra "1" to vector_AB and transpose it to get size (4 x 1)
vector_AB(4) = 1;
vector_AB = vector_AB';

Kx = vector_K(1);
Ky = vector_K(2);
Kz = vector_K(3);
V0 = 1 - cos(angle);
C = cos(angle);
S = sin(angle);

RotMat = [Kx*Kx*V0 + C, Kx*Ky*V0 - Kz*S, Kx*Kz*V0 + Ky*S;
          Kx*Ky*V0 + Kz*S, Ky*Ky*V0 + C, Ky*Kz*V0 - Kx*S;
          Kx*Kz*V0 - Ky*S, Ky*Kz*V0 + Kx*S, Kz*Kz*V0 + C;
          0, 0, 0];

TM = [RotMat, vector_AB];      


end

